### ISSUE TITLE
Gamemode: Horde Mode

### ISSUE THREAD

**Issue by Prolonpo** [AoTTG-2/AoTTG-2#293]
**How will this feature improve the game?**

Will enhance the gameplay for AoTTG 2, create more memories and new gameplays. Would allow the game to have more novelties so that players would get engage for longer.

**Describe the solution you'd like**

Objective:
• Defend the gates - titans attack the gates which have their own HP, when the gates' HP drops to 0, the gates breaks and players need to retreat to the inner wall
• Survive for a set amount of waves (to be determined)
• There are three sets of walls with their protruding districts, the players lose when the titan break the innermost wall.

Can use cannons, 3DMG, Thunder Spears (when that becomes available) to kill the titans. However, cannon balls, gas for the 3DMG, blades, and thunder spears will cost points. In addition, there is limited ammunition for cannon balls and thunder spears - currently thinking of 10 cannon balls and 4 thunder spears which one player can have at once.

Killing titans & Striking Titans at certain spots (Eyes, Ankles) will allow the player to earn points that they can spend on buying the items mentioned above. Killing titans will grant significantly more points than striking the Titans' Eyes and Ankles.

Players can give items by dropping them on the ground for other players to pick them up.

Players who have been punched or slapped by titans will not automatically die. They will have a 1 min time frame where another player can revive them.

Players who have been eaten die automatically.

Dead players will respawn during the preparation phase of the next wave.

Signal flares will have a significant use in this game mode:
• Green flares would highlight all players in a certain distance of where the flare has been shot
• Red flares would highlight all titans in a certain radius of where the flare has been shot
• Black flares would highlight dangerous titans in a certain radius of where the flare has been shot
• Emergency flare (blue or magenta, I forgot) will pin the shooter's position. Can be used by downed players to request revive
• Yellow flare will pin the location resupply stations in a certain radius of where the flare has been shot.

**Describe alternatives you've considered**

N/A.

**Implementation**
@AoTTG-2/development-team 
- [ ] Confirm #114 and #160 are done. Implementation may also start once #160 is done, but the item concept would be postponed until #114 is introduced. Discuss with @Lithrun 
- [ ] Extend GamemodeBase with a new Gamemode named HordeGamemode
- [ ] Create prefabs and scripts to determine gate logic
- [ ] Introduce a new titan behavior that overrides the OnWandering and OnChasing behavior so that titans prioritize on the gates.
- [ ] Investigate on how to implement points and a shop to add inventory items

@AoTTG-2/mapping-team 
- [ ] Create a map which can support this gamemode. Should have multiple walls and gates.

**Comment by Lithrun** [comment]
Sounds like a great new gamemode for AoTTG2. Thank you for the suggestion, and will be moving this to the future milestone. Added implementation details in the issue 👍

**Comment by Prolonpo** [comment]
From @CriminalTBG

• Executioner from Hell Trost Survival Map
• A wave based map where you can use the executioner from hell in certain conditions.
• Players are supposed to keep the titans away from the district, but if they get too close, they can activate the executioner by activating certain levers on top of the wall (the cost of resetting it is 30 seconds.) The executioner then kills all the titans below it, but the gate itself has a health, and is not restorable once it decreases.

The players are expected to defend for x waves in this map

Victory Conditions:
              -The gate hasn't been breached
              -At least one player is alive
              -All of the titans have been killed
The supplies are located inside the city, most of the city is visible, but either inaccessible or just useless

Reference post: https://discord.com/channels/681641241125060652/683376807101071489/761906450762956830

Also attached a conversation between @CriminalTBG and @Lithrun about this implementation to Horde Mode.
![Ekran_goruntusu_2020-11-19_213755](https://user-images.githubusercontent.com/72053473/99709516-486f7d00-2a6d-11eb-8db5-61ebaa146257.png)
